### PR TITLE
Octree impl

### PR DESCRIPTION
**Pull request by bbeck13**
https://github.com/bbeck13/RhythmRunner/issues/48
There is some extra stuff that accidentally got in here but all the functional changes have to do with the implementation of the octree see [https://en.wikipedia.org/wiki/Octree](https://en.wikipedia.org/wiki/Octree) for reference. If you're confused about what some of this logic does.

### CODE REVIEW

File: src/game_renderer/Shape.cpp [hunk 1]
```diff
@@ -13,191 +13,190 @@ Shape::Shape() : eleBufID(0), posBufID(0), norBufID(0), texBufID(0), vaoID(0) {}
...
+  bool rc = tinyobj::LoadObj(shapes, objMaterials, errStr, meshName.c_str());
+  if (!rc) {
+    std::cerr << errStr << std::endl;
+  } else {
+    posBuf = shapes[0].mesh.positions;
+    norBuf = shapes[0].mesh.normals;
+    texBuf = shapes[0].mesh.texcoords;
+    eleBuf = shapes[0].mesh.indices;
+    Normalize();
+  }
```
Comment (josepharhar) [inline src/game_renderer/Shape.cpp:32]: Are all the changes in this file just clang-format or is there anything functional?

File: src/game_state/Octree.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  return objects;
+}
+
+Node* Octree::constructOctree(std::vector<std::shared_ptr<GameObject>>* objs) {
+  // leaf
+  if (objs->size() <= OBJS_IN_LEAF) {
+    glm::vec3 min, max;
+    if (objs->empty()) {
+      std::cerr << "Catastrophic system failure. Unrecoverable" << std::endl;
+      exit(EXIT_FAILURE);
```
Comment (josepharhar) [inline src/game_state/Octree.cpp:33]: lol :ok_hand:

File: src/game_state/Octree.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }
+    return new Node(objs, AxisAlignedBox(getMin(objs), getMax(objs)),
+                    new std::vector<Node*>());
```
Comment (josepharhar) [inline src/game_state/Octree.cpp:36]: Any reason to use new instead of shared_ptr?

File: src/game_state/GameObject.h [hunk 4]
```diff
@@ -19,6 +19,8 @@ class GameObject {
   virtual ~GameObject();
 
   virtual std::shared_ptr<Shape> GetModel() = 0;
+  virtual bool isCollectible() = 0;
+  virtual bool isObstacle() = 0;
```
Comment (josepharhar) [inline src/game_state/GameObject.h]: These should be replaced with a single method that returns an enum or enum class

File: src/game_state/MovingObject.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+class MovingObject {
+ public:
+
+  virtual float GetVelocity() {
+    return velocity;
+  }
+
+  virtual std::shared_ptr<std::vector<glm::vec3>> GetPath() {
+    return path;
+  }
```
Comment (josepharhar) [inline src/game_state/MovingObject.h]: GetPath() confuses me. What do the vec3s represent?

File: src/game_renderer/GameRenderer.cpp [hunk 6]
```diff
@@ -133,12 +135,17 @@ void GameRenderer::Render(std::shared_ptr<GameState> game_state) {
...
 
-  for (std::shared_ptr<Platform> platform : *level->getPlatforms()) {
-    if (!ViewFrustumCulling::IsCulled(platform->GetBoundingBox(), vfplane)) {
-      MV = platform->GetTransform();
-      glUniformMatrix4fv(current_program->getUniform("MV"), 1, GL_FALSE,
-                         glm::value_ptr(MV.topMatrix()));
-      platform->GetModel()->draw(current_program);
+  for (std::shared_ptr<GameObject> obj : *level->getObjects()) {
+    if (obj->isObstacle()) {
+      if (Platform* platform = dynamic_cast<Platform*>(obj.get())) {
```
Comment (josepharhar) [inline src/game_renderer/GameRenderer.cpp]: Since you already know the object type, and to avoid using raw pointers, I suggest using a std::static_pointer_cast to Obstacle instead of dynamic_cast.

You could also combine the note rendering in this loop so that you don't have to iterate level->getObjects() twice

File: src/game_renderer/GameRenderer.cpp [hunk 7]
```diff
@@ -151,13 +158,17 @@ void GameRenderer::Render(std::shared_ptr<GameState> game_state) {
...
-  for (std::shared_ptr<Note> note : *level->getNotes()) {
-    if (!ViewFrustumCulling::IsCulled(note->GetBoundingBox(), vfplane) &&
-        !note->GetCollected()) {
-      MV = note->GetTransform();
-      glUniformMatrix4fv(current_program->getUniform("MV"), 1, GL_FALSE,
-                         glm::value_ptr(MV.topMatrix()));
-      note->GetModel()->draw(current_program);
+  for (std::shared_ptr<GameObject> obj : *level->getObjects()) {
+    if (obj->isCollectible()) {
+      if (Note* note = dynamic_cast<Note*>(obj.get())) {
```
Comment (josepharhar) [inline src/game_renderer/GameRenderer.cpp]: dynamic_cast -> std::static_pointer_cast

File: src/game_updater/GameUpdater.cpp [hunk 8]
```diff
@@ -49,6 +62,25 @@ void GameUpdater::Update(std::shared_ptr<GameState> game_state) {
...
+  // reset the player
+  game_state->SetDone(false);
+  game_state->GetPlayer()->SetPosition(Player::INITIAL_POSITION);
+  game_state->GetPlayer()->SetScore(0);
+
+  // reset collectibles
+  for (std::shared_ptr<GameObject> obj :
+       *game_state->GetLevel()->getObjects()) {
+    if (obj->isCollectible()) {
+      if (Collectible* c = dynamic_cast<Collectible*>(obj.get())) {
```
Comment (josepharhar) [inline src/game_updater/GameUpdater.cpp]: dynamic_cast -> std::static_pointer_cast

File: src/game_updater/GameUpdater.cpp [hunk 9]
```diff
@@ -111,21 +129,23 @@ void GameUpdater::UpdatePlayer(std::shared_ptr<GameState> game_state) {
...
-      game_state->GetPlayer()->SetScore(game_state->GetPlayer()->GetScore() +
-                                        1);
+  // collide!
+  std::shared_ptr<std::unordered_set<std::shared_ptr<GameObject>>>
+      collidingObjects = GetCollidingObjects(player->GetBoundingBox(),
+                                             game_state->GetLevel()->getTree());
+  std::vector<std::shared_ptr<GameObject>> colliding_platforms;
+  for (std::shared_ptr<GameObject> g : *collidingObjects) {
+    if (g->isCollectible()) {
+      if (Collectible* c = dynamic_cast<Collectible*>(g.get())) {
```
Comment (josepharhar) [inline src/game_updater/GameUpdater.cpp]: dynamic_cast -> std::static_pointer_cast

**Review by josepharhar — COMMENTED** [review]
In addition to these comments, there was a lot of places where functions were implemented in headers that I didn't point out

File: src/game_renderer/Shape.cpp (hunk 1 above)
Comment (bbeck13) [reply src/game_renderer/Shape.cpp:32]: none of it is functional I must have a replaced the file when debugging

File: src/game_state/Octree.cpp (hunk 3 above)
Comment (bbeck13) [reply src/game_state/Octree.cpp:36]: it's faster, I noticed a performance drop when using shared pointers here

File: src/game_state/MovingObject.h (hunk 5 above)
Comment (bbeck13) [reply src/game_state/MovingObject.h]: This wasn't supposed to be on this branch but the vec3s are points that the object travels between forming a circuit

File: src/game_renderer/GameRenderer.cpp (hunk 6 above)
Comment (bbeck13) [reply src/game_renderer/GameRenderer.cpp]: needs to be std::dynamic_pointer_cast because here we need the specific type of the object (Platform in this case) to determine rendering information. That's also why we can't combine the rendering loops. because we need to bind a different program of each type of object. We already have culling in place so looping though all the objects in view a couple times during the render phase shouldn't be a problem
Alex is also looking into speeding up the drawing process
https://github.com/bbeck13/RhythmRunner/issues/58

**Comment by bbeck13** [comment]
I changed the view frustum culling code to use the octree to determine what objects are culled. and revised the code per @josepharhar's comments

File: src/game_state/Octree.cpp (hunk 3 above)
Comment (josepharhar) [reply src/game_state/Octree.cpp:36]: Oh wow really? Maybe we should reconsider shared_ptr in some other places. How big was the performance hit? Like 10 fps or something?

Comment (bbeck13) [reply src/game_state/Octree.cpp:36]: ~15, I was going to refactor some of the other code that uses shared pointers but this does the job and I'm back to ~55 fps

File: src/game_renderer/GameRenderer.cpp (hunk 6 above)
Comment (josepharhar) [reply src/game_renderer/GameRenderer.cpp]: If we are going with dynamic_casting, then what's the point of having/using isObstacle()/GetType() here? The if() check around the dynamic cast should take care of it, right?

File: src/game_state/Octree.cpp (hunk 3 above)
Comment (josepharhar) [reply src/game_state/Octree.cpp:36]: I'm going to make an issue for investigating replacing shared_ptrs in some places. We should have a group conversation about how we could change the memory management for GameObjects, specifically if we can say whether or not their lifetime will persist for the entire length of the game. Do you know specifically how shared_ptrs with Nodes could have slowed it down? Was make_shared being called every frame? Or were there a lot of shared_ptr objects being copied around through function calls each frame?

Comment (bbeck13) [reply src/game_state/Octree.cpp:36]: I'm using a queue to traverse through the nodes and each node had shared pointers to at most 8 sub nodes so I think enqueuing shared_ptr to the node caused for a lot of plumbing on the shared_ptrs part.

File: src/game_renderer/GameRenderer.cpp (hunk 6 above)
Comment (bbeck13) [reply src/game_renderer/GameRenderer.cpp]: https://www.youtube.com/watch?v=t1WSD_cnRbA
(Seriously though I'll remove it)

File: src/RhythmRunner.cpp [hunk 10]
```diff
@@ -25,13 +25,15 @@ int main(int argc, char** argv) {
...
 
+  std::shared_ptr<Level> level = levelGenerator->generateLevel();
   std::shared_ptr<GameState> game_state = std::make_shared<GameState>(
       level, std::make_shared<GameCamera>(), std::make_shared<Player>());
-
-  std::shared_ptr<GameRenderer> renderer = std::make_shared<GameRenderer>();
-  renderer->Init(ASSET_DIR, game_state);
+  // had to separate this since gl initialization needs to be done before
+  // generating the level
+  renderer->InitState(game_state);
```
Comment (josepharhar) [inline src/RhythmRunner.cpp:34]: It looks like InitState() just sets up the keybindings. Why does this need to happen after generating the level? And why does generating the level require gl initialization?

Comment (bbeck13) [reply src/RhythmRunner.cpp:34]: The bounding boxes require information from the shape which needs to gl to be initialized and I need the bounding boxes to set up the octree which is a part of the level.

**Comment by Matthew-Stewart** [comment]
QA ⭕️

File: src/game_renderer/Shape.cpp [hunk 11]
```diff
@@ -13,191 +13,190 @@ Shape::Shape() : eleBufID(0), posBufID(0), norBufID(0), texBufID(0), vaoID(0) {}
...
+    texBuf[n * 3 + 1] = v;
+  }
 }
 
 void Shape::Normalize() {
-   float minX, minY, minZ;
-   float maxX, maxY, maxZ;
-   float scaleX, scaleY, scaleZ;
-   float shiftX, shiftY, shiftZ;
-   float epsilon = 0.001;
```
Comment (josepharhar) [inline src/game_renderer/Shape.cpp:51]: epsilon should be put back in

**Comment by bbeck13** [comment]
Rebased on master made octree faster (added easily configurable option for quadtree) added platform stuffs

File: src/game_state/GameObject.cpp [hunk 12]
```diff
@@ -32,20 +32,30 @@ glm::vec3 GameObject::GetScale() {
...
-  AxisAlignedBox box(GetModel(), GetTransform().topMatrix());
-  return box;
+  return AxisAlignedBox(GetModel(), GetTransform().topMatrix());
+}
+
+AxisAlignedBox GameObject::GetBoundingBox(std::shared_ptr<Shape> model,
+                                          glm::vec3 scale,
+                                          glm::vec3 position) {
+  // TODO(jarhar): make this more efficient by caching calculated box
+  return AxisAlignedBox(model, GetTransform(position, scale).topMatrix());
```
Comment (josepharhar) [inline src/game_state/GameObject.cpp]: This static method seems more like a constructor for an AxisAlignedBox than anything. You should probably move it to AxisAlignedBox

File: src/game_updater/GameUpdater.cpp [hunk 13]
```diff
@@ -100,69 +136,76 @@ void GameUpdater::UpdatePlayer(std::shared_ptr<GameState> game_state) {
...
+    if (std::shared_ptr<MovingObject> moving =
+            std::dynamic_pointer_cast<MovingObject>(player->GetGround())) {
+      // Move player with moving object
+      glm::vec3 movementVector = moving->GetMovementVector();
+      player->SetVerticalVelocity(movementVector.y * moving->GetVelocity().y);
+      player->SetZVelocity(movementVector.z * moving->GetVelocity().z);
+      collision_width += movementVector.y * moving->GetVelocity().y;
+    } else {
+      // player is stuck to ground, make sure velocity is neutralized
+      player->SetVerticalVelocity(0.0f);
```
Comment (josepharhar) [inline src/game_updater/GameUpdater.cpp]: do you think ZVelocity should be set here too?

File: src/game_updater/GameUpdater.cpp [hunk 14]
```diff
@@ -100,69 +136,76 @@ void GameUpdater::UpdatePlayer(std::shared_ptr<GameState> game_state) {
...
+    float colliding_max_y = colliding_object->GetBoundingBox().GetMax().y;
+    AxisAlignedBox player_box = player->GetBoundingBox();
+    float prev_player_min_y = previous_player_box.GetMin().y;
+
+    // If the player's bounding box was "above" (within 1 velocity unit +
+    // allowed COLLISION_WIDTH error) the platform's box before
+    // collision,
+    // then we consider the collision "landing" on the platform
+    if (std::abs(prev_player_min_y - colliding_max_y) <
+        (std::abs(player->GetVerticalVelocity()) + collision_width)) {
```
Comment (josepharhar) [inline src/game_updater/GameUpdater.cpp]: I don't fully understand the case in which this type of check is needed, but I believe it :ok_hand:

File: src/game_updater/GameUpdater.cpp [hunk 15]
```diff
@@ -100,69 +136,76 @@ void GameUpdater::UpdatePlayer(std::shared_ptr<GameState> game_state) {
...
-    // TODO(jarhar): further consider what is happening when we are colliding
-    // with two objects
-    Reset(game_state);
-  } else if (colliding_platforms.size() == 0) {
-    // No collision
-  } else {
-    // Handle the collision
-    std::shared_ptr<GameObject> colliding_object = colliding_platforms[0];
+  // Handle the collisions
+  for (std::shared_ptr<GameObject> colliding_object : colliding_platforms) {
```
Comment (josepharhar) [inline src/game_updater/GameUpdater.cpp]: Does looping through colliding platforms like this seem to work well? I would think that it would be weird if you get snapped to two different ground platforms based on which comes last in this list, but if it works then I'm ok with it

File: src/game_state/Player.h [hunk 16]
```diff
@@ -17,16 +17,20 @@ class Player : public GameObject {
...
 
+  void Reset();
   float GetVerticalVelocity();
+  float GetZVelocity();
   bool GetSpacebarDown();
   std::shared_ptr<GameObject> GetGround(); // null if no ground
   std::shared_ptr<Shape> GetModel() override;
+  ObjectType GetType() override;
 
   void SetVerticalVelocity(float vertical_velocity);
```
Comment (josepharhar) [inline src/game_state/Player.h]: Since you added ZVelocity, you should change the naming of VerticalVelocity to YVelocity in the getter and setter to make it consistent

File: src/game_updater/GameUpdater.cpp (hunk 13 above)
Comment (bbeck13) [reply src/game_updater/GameUpdater.cpp]: thanks for catching that

File: src/game_state/GameObject.cpp (hunk 12 above)
Comment (bbeck13) [reply src/game_state/GameObject.cpp]: will do

File: src/game_updater/GameUpdater.cpp (hunk 15 above)
Comment (bbeck13) [reply src/game_updater/GameUpdater.cpp]: If we get snapped to two different grounds It doesn't matter since if one is higher than the other or if one is coming up from underneath we will collide with it in the next pass and reset our ground. This logic is a bit confusing though I'll clean it up

File: src/game_updater/GameUpdater.cpp (hunk 14 above)
Comment (bbeck13) [reply src/game_updater/GameUpdater.cpp]: I noticed that if the player falls for a long time onto a platform their bounding box moves below the colliding box before the game can detect a collision, this adds some tolerance for that based on the players vertical velocity.

tl;dr
it's so the player doesn't take falling damage

File: src/game_state/Player.h (hunk 16 above)
Comment (bbeck13) [reply src/game_state/Player.h]: 👍

**Comment by josepharhar** [comment]
CR :beer: QA :beers:
Thanks for implementing my suggestions, this has been a big code review

**Comment by bbeck13** [comment]
thanks for reviewing it :1st_place_medal:
